plugins/jobtype/src/azkaban/jobtype/HadoopTuningSecurePigWrapper.java

```diff
@@ -0,0 +1,181 @@
+  private static int jobTryCount = 1;
+  /**
+   * Is job failed because of tuning parameters
+   */
+  private static boolean isTuningError = false;
+
+  private static final String TUNING_JOB_RETRY_COUNT = "tuning.job.retry.count";
+
+  static {
+    logger = Logger.getRootLogger();
+  }
+
+  /**
+   * This function runs the pig job with tuning enabled. In case tuning job fails in first try, it asks for
+   * last best parameter from tuning and re-run the job. In case tuning end point is not responding this job
+   * will run with default parameters.
+   * @param args
+   * @throws Throwable
+   */
+  public static void main(final String[] args) throws Exception {
+    Properties jobProps = HadoopSecureWrapperUtils.loadAzkabanProps();
+    Props initialJobprops = new Props(null, jobProps);
+    boolean retry = false;
+    boolean firstTry = true;
+    if (props.containsKey(TUNING_JOB_RETRY_COUNT)) {
+      maxJobRetry = props.getInt(TUNING_JOB_RETRY_COUNT);
+    }
+
+    while (jobTryCount <= maxJobRetry && (retry || firstTry)) {
+      jobTryCount++;
+      firstTry = false;
+      props = Props.clone(initialJobprops);
+      props.put(TuningCommonConstants.AUTO_TUNING_RETRY, retry + "");
+
+      TuningParameterUtils.updateAutoTuningParameters(props);
+
+      HadoopTuningConfigurationInjector.prepareResourcesToInject(props,
+          HadoopTuningSecurePigWrapper.getWorkingDirectory(props));
+
+      HadoopTuningConfigurationInjector.injectResources(props);
+
+      // special feature of secure pig wrapper: we will append the pig error file
+      // onto system out
+      pigLogFile = new File(System.getenv("PIG_LOG_FILE"));
+
+      try {
+        if (HadoopSecureWrapperUtils.shouldProxy(jobProps)) {
+          String tokenFile = System.getenv(HADOOP_TOKEN_FILE_LOCATION);
+          UserGroupInformation proxyUser = HadoopSecureWrapperUtils.setupProxyUser(jobProps, tokenFile, logger);
+          proxyUser.doAs(new PrivilegedExceptionAction<Void>() {
+            @Override
+            public Void run() throws Exception {
+              runPigJob(args);
+              return null;
+            }
+          });
+        } else {
+          runPigJob(args);
+        }
+      } catch (Exception t) {
+        retry = false;
+        System.out.println("Error " + isTuningError + ", tryCount:" + jobTryCount + ", maxRetry:" + maxJobRetry);
+        if (isTuningError && jobTryCount <= maxJobRetry) {
+          System.out.println("Error due to auto tuning parameters ");
+          retry = true;
+        } else {
+          throw t;
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("deprecation")
+  public static void runPigJob(String[] args) throws Exception {
+    PigStats stats = PigUtil.runPigJob(args, props);
+
+    PigUtil.dumpHadoopCounters(stats, props);
+
+    if (stats.isSuccessful()) {
+      return;
+    }
+
+    if (pigLogFile != null) {
+      handleError(pigLogFile);
+    }
+
+    if (isTuningError && jobTryCount <= maxJobRetry) {
```

Comment:
I don't really like the use of a static `isTuningError` here. I think it would be cleaner to have `handleError` and `runPigJob` return boolean values indicating whether or not there was a tuning error, or to create a new exception type `TuningException` which is thrown to indicate tuning-specific errors.

Reply:
Removed static isTuningError and added TuningException.